File: app/src/main/java/com/nathaniel/motus/cavevin/model/Cell.java
package com.nathaniel.motus.cavevin.model;

import android.util.Log;

import java.util.ArrayList;

public class Cell {
    //a cell is a place where you can store a bottle with a comment, a quantity, an origin and a statut
    //a cellar is an ArrayList of cells

    private static final String SEPARATOR="|";

    //**********************************************************************************************
    //The collection of cells

    //List of the cells
    private static ArrayList<Cell> sCellPool=new ArrayList<>();
    //**********************************************************************************************

    //the bottle
    private Bottle mBottle;

    //how the bottle was acquired
    private String mOrigin;

    //Stock
    private int mStock;

    //Comment proper to the cellar
    private String mCellComment;

    //Flag expanded, for recyclerview - is not saved in database
    public boolean isExpanded=false;

    //**********************************************************************************************
    //Getters and setters
    //**********************************************************************************************

    public Bottle getBottle() {
        return mBottle;
    }

    public void setBottle(Bottle bottle) {
        mBottle = bottle;
    }

    public String getOrigin() {
        return mOrigin;
    }

    public void setOrigin(String origin) {
        mOrigin = origin;
    }

    public int getStock() {
        return mStock;
    }

    public void setStock(int stock) {
        mStock = stock;
    }

    public String getCellComment() {
        return mCellComment;
    }

    public void setCellComment(String cellComment) {
        mCellComment = cellComment;
    }

    public static int getNumberOfCells() {
        return sCellPool.size();
    }

    public static ArrayList<Cell> getCellPool() {
        return sCellPool;
    }

    //**********************************************************************************************
    //Constructors
    //**********************************************************************************************

    public Cell(Bottle bottle, String origin, int stock, String cellComment,Boolean isReferenced) {
        //Creates a cell which is reference in the CellPool is isReferenced
        mBottle = bottle;
        mOrigin = origin;
        mStock = stock;
        mCellComment = cellComment;

        if(isReferenced) {
            sCellPool.add(this);
        }
    }

    public Cell(Boolean isReferenced) {
        //Creates a cell
        //if isReferenced is true, it is referenced in the CellPool
        //if isReferenced is false, it is not
        mBottle=Bottle.getBottleCatalog().get(0);
        mOrigin="";
        mStock=0;
        mCellComment="";

        if(isReferenced) {
            sCellPool.add(this);
        }
    }

    //**********************************************************************************************
    //Modifiers
    //**********************************************************************************************

    public void clearCell(){
        //clear a cell of its content

        mBottle=Bottle.getBottleCatalog().get(0);
        mOrigin="";
        mStock=0;
        mCellComment="";
    }

    public void setCellParametersOf(Cell cell){
        //copy the parameters of cell to this

        this.setBottle(cell.getBottle());
        this.setOrigin(cell.getOrigin());
        this.setStock(cell.getStock());
        this.setCellComment(cell.getCellComment());
    }

    public void removeCell(){
        //remove cell from cellars and cell pool
        //if bottle of this has no use case, remove it from catalog

        Bottle bottle=this.getBottle();

        if (this.findUseCaseCellar(0)!=null) this.findUseCaseCellar(0).getCellList().remove(this);
        sCellPool.remove(this);

        if (bottle.findUseCaseCell(0) == null) bottle.removeBottleFromCatalog();
    }

    public static void clearCellPool(){
        //empty cell pool

        sCellPool.clear();
    }

    //**********************************************************************************************
    //Manipulators
    //**********************************************************************************************

    public Boolean isBottleUseCase(Bottle bottle){
        //returns true if the bottle is used in this cell

        if (mBottle==bottle) return true;
        else return false;
    }

    public Cellar findUseCaseCellar(int fromIndex){
        //find the first use case cellar of this

        if (fromIndex>=Cellar.getCellarPool().size()) return null;
        else{
            if (Cellar.getCellarPool().get(fromIndex).isCellUseCase(this)) return Cellar.getCellarPool().get(fromIndex);
            else return findUseCaseCellar(fromIndex+1);
        }
    }

}

File: app/src/main/java/com/nathaniel/motus/cavevin/model/Cell.kt
package com.nathaniel.motus.cavevin.model

import com.nathaniel.motus.cavevin.model.Bottle
import com.nathaniel.motus.cavevin.model.Cellar
import com.nathaniel.motus.cavevin.model.JsonObject
import com.nathaniel.motus.cavevin.model.CellComparator
import java.util.ArrayList

class Cell {
    //**********************************************************************************************
    //Getters and setters
    //**********************************************************************************************
    //**********************************************************************************************
    //the bottle
    var bottle: Bottle

    //how the bottle was acquired
    var origin: String

    //Stock
    var stock: Int

    //Comment proper to the cellar
    var cellComment: String

    //Flag expanded, for recyclerview - is not saved in database
    @JvmField
    var isExpanded = false

    //**********************************************************************************************
    //Constructors
    //**********************************************************************************************
    constructor(
        bottle: Bottle,
        origin: String,
        stock: Int,
        cellComment: String,
        isReferenced: Boolean
    ) {
        //Creates a cell which is reference in the CellPool is isReferenced
        this.bottle = bottle
        this.origin = origin
        this.stock = stock
        this.cellComment = cellComment
        if (isReferenced) {
            cellPool.add(this)
        }
    }

    constructor(isReferenced: Boolean) {
        //Creates a cell
        //if isReferenced is true, it is referenced in the CellPool
        //if isReferenced is false, it is not
        bottle = Bottle.Companion.bottleCatalog.get(0)
        origin = ""
        stock = 0
        cellComment = ""
        if (isReferenced) {
            cellPool.add(this)
        }
    }

    //**********************************************************************************************
    //Modifiers
    //**********************************************************************************************
    fun clearCell() {
        //clear a cell of its content
        bottle = Bottle.Companion.bottleCatalog.get(0)
        origin = ""
        stock = 0
        cellComment = ""
    }

    fun setCellParametersOf(cell: Cell) {
        //copy the parameters of cell to this
        bottle = cell.bottle
        origin = cell.origin
        stock = cell.stock
        cellComment = cell.cellComment
    }

    fun removeCell() {
        //remove cell from cellars and cell pool
        //if bottle of this has no use case, remove it from catalog
        val bottle = bottle
        if (findUseCaseCellar(0) != null) findUseCaseCellar(0)?.cellList?.remove(this)
        cellPool.remove(this)
        if (bottle.findUseCaseCell(0) == null) bottle.removeBottleFromCatalog()
    }

    //**********************************************************************************************
    //Manipulators
    //**********************************************************************************************
    fun isBottleUseCase(bottle: Bottle): Boolean {
        //returns true if the bottle is used in this cell
        return if (this.bottle === bottle) true else false
    }

    fun findUseCaseCellar(fromIndex: Int): Cellar? {
        //find the first use case cellar of this
        return if (fromIndex >= Cellar.Companion.cellarPool.size) null else {
            if (Cellar.Companion.cellarPool.get(fromIndex)
                    .isCellUseCase(this)
            ) Cellar.Companion.cellarPool.get(fromIndex) else findUseCaseCellar(fromIndex + 1)
        }
    }

    companion object {
        //a cell is a place where you can store a bottle with a comment, a quantity, an origin and a statut
        //a cellar is an ArrayList of cells
        private const val SEPARATOR = "|"

        //**********************************************************************************************
        //The collection of cells
        //List of the cells
        val cellPool = ArrayList<Cell>()
        val numberOfCells: Int
            get() = cellPool.size

        fun clearCellPool() {
            //empty cell pool
            cellPool.clear()
        }
    }
}